![](https://www.dii.uchile.cl/wp-content/uploads/2021/06/Magi%CC%81ster-en-Ciencia-de-Datos.png)

# Proyecto: Innovación Tecnológica en Bodoque Bank

**MDS7202: Laboratorio de Programación Científica para Ciencia de Datos**

### Cuerpo Docente:

- Profesor: Pablo Badilla, Ignacio Meza De La Jara
- Auxiliar: Sebastián Tinoco
- Ayudante: Diego Cortez M., Felipe Arias T.

*Por favor, lean detalladamente las instrucciones de la tarea antes de empezar a escribir.*


----

## Reglas

- Fecha de entrega: 01/06/2021
- **Grupos de 2 personas.**
- Cualquier duda fuera del horario de clases al foro. Mensajes al equipo docente serán respondidos por este medio.
- Estrictamente prohibida la copia. 
- Pueden usar cualquier material del curso que estimen conveniente.

# Motivación

![](https://www.geekmi.news/__export/1621969098810/sites/debate/img/2021/05/25/juan-carlos-bodoque-1.jpg_554688468.jpg)

Juan Carlos Bodoque, el famoso periodista y empresario, decidió diversificar su portafolio de negocios y crear su propio banco. Después de varios años de investigar y analizar el mercado financiero, finalmente logró fundar su entidad bancaria con el objetivo de ofrecer a sus clientes una experiencia personalizada y confiable en sus transacciones financieras. 

Sin embargo, con las nuevas tecnologías, aparecen nuevos desafíos para la joven entidad bancaria. Por ello, Bodoque decide invertir en un equipo de expertos en tecnología y finanzas, para que Bodoque Bank implemente las últimas innovaciones en seguridad y servicio al cliente para garantizar la satisfacción y fidelización de sus clientes.

El primer objetivo de la entidad bancaria será la detección de potenciales clientes fraudulentos, para ello Bodoque Bank les hace entrega de un extenso dataset en el que se registran las actividades que han realizado sus clientes durante los últimos meses. Uno de los puntos que resaltan al pasar el conjunto de datos es que el nombre de los usuarios está protegido y que consideren cada una de las filas como una muestra independiente de la otra.

Tras la solicitud, uno de los mayores accionistas del banco llamado Mario Hugo, les sugiere que al momento de realizar el proyecto tomen las siguientes consideraciones:

- Realicen un análisis exhaustivo de cada variable, ya que los directores tienden a ser quisquillosos con puntos o definiciones sin algún fundamento claro.
- Deben ser muy claros de cómo encontrar los outliers y si es posible expliquen cómo están determinando que un cliente es fraudulento.
- Es muy probable que existan clientes con diferentes comportamientos bancarios, por lo que sería muy buena decisión de separar a los clientes en grupos etarios.


### Definición Formal del Problema

El dataset con el que se trabajará en este primer proyecto será [_The Bank Account Fraud (BAF)_](https://www.kaggle.com/datasets/sgpjesus/bank-account-fraud-dataset-neurips-2022). Este consiste en un dataset para evaluar métodos de detección de fraudes bancarios, el cuál según sus autores es:

- Realista: Se basa en un actualizado dataset de casos reales de detección de fraude bancario.
- Sesgado: Poseen distintos tipos de bias.
- Desbalanceado: Clase positiva extremadamente pequeña.
- Dinámico: Presenta datos temporales y cambios de distribución.
- Preserva la privacidad: Protege a los potenciales clientes a través de la aplicación de privacidad diferencial.

El proyecto tiene por objetivo evaluar los conocimientos adquiridos en la primera mitad del curso, consistente en manejo de datos tabulares (I/O, manipulación, agregaciones, merge y visualizaciones) más la primera parte de modelos consistente en preprocesamiento de datos y detección de anomalías. Por ende, el proyecto consiste en dos tareas principales:

1. Generar un Análisis exploratorio de Datos (EDA) que describa completamente el dataset.
2. Buscar anomalías de forma automatizada.

**Importante:** Esta permitido el uso de librerías externas a las vistas en clases para profundizar aún mas en los análisis. Sin embargo, al momento de utilizar cualqueir metodo deberán explicar que hace y el porque de su aplicación.


---

## Secciones requeridas en la entrega

La siguiente lista detalla las secciones que debe contener su notebook para resolver el proyecto. Es importante que al momento de desarrollar cada una de las secciones, estas sean escritas en un formato tipo **informe**, donde describan detalladamente cada uno de los puntos realizados.

### 1. Introducción [0.5 Puntos]

Elaborar una breve introducción con todo lo necesario para entender qué realizarán durante su proyecto. La idea es que describan de manera formal el proyecto con sus propias palabras y logren describir algunos aspectos básicos tanto del dataset como del análisis a realizar sobre los datos.

Por lo anterior, en esta sección ustedes deberán ser capaces de:

- Describir la tarea asociada al dataset.
- Describir brevemente los datos de entrada que les provee el problema.
- Plantear hipótesis de cómo podrían abordar el problema.

### 2. Lectura y Manejo de Datos [1 Punto]

Este punto tiene por objetivo evaluar la capacidad de leer y manejar distintas fuentes de datos y unirlas.
Para esto, se les provee de 3 datasets distintos: 

- `df_1.parquet` que contiene una cierta cantidad de ejemplos.
- `df_2.parquet` que contiene el resto de los ejemplos.
- `df_email_phone.parquet` que contiene características exclusivas relacionadas a los emails y teléfonos de los ejemplos.

Tanto `df_1` como `df_2` son dataframes que contienen conjuntos (no necesariamente disjuntos) de ejemplos obtenidos a partir de distintas fuentes (la primera más confiable que la segunda)  mientras que `df_email_phone` contiene nuevas características que deben ser fusionadas con el conjunto de ejemplos.
Dicho esto, para este punto se le solicita:

1. Cargar los datasets usando pandas.
2. Explorar superficialmente los datsets por medio de `.head(5)` e `.info()` y reportar número de filas y columnas de cada dataset.
3. Unir los datasets con los conjuntos de ejemplos en un único dataset.
4. Combinar el datasets con ejemplos con las nuevas características usando la estrategia `outer`.
5. Explorar nulos y duplicados de datos (i.e., generar conteos de cada uno y ver algunos de sus ejemplos), explicar su origen y entregar una solución sencilla a estos casos.
6. Limpiar datos y reportar modificaciones a los datasets originales (cuántos ejemplos se eliminaron, cuántos ejemplos quedaron finalmente, etc...).
7. Realizar una segmentación etaria de los datos creando una columna llamada `segmentacion_etaria` para clasificar a los diferentes usuarios. Para la segmentación utilice los siguientes rangos:
   - Joven (Menor a 18 años)
   - Adulto-Joven (18 - 26 años)
   - Adulto (27- 59 años)
   - Persona Mayor (60 años o mas).


### 3. Análisis Exploratorio de Datos [1 Punto]

Esta sección se busca evaluar la capacidad de generar reportes y conclusiones a partir de los patrones observados en estos.

- Utilizando `pandas-profiling` (nombre de la librería: `ydata-profiling`), explore:

    * Analizar cantidad de datos nulos, tipos de datos, duplicados, distribuciones de las variables a través de histogramas.
    * Generar visualizaciones de las interacciones (como por ejemplo, una scatter matrix) en las distintas variables.
    * Ver las correlaciones entre las distintas variables y los valores faltantes de cada una de estas. 
    * Reportar los patrones interesantes observados. Ejemplos:
        - Se observó que las variables numéricas A, B y C siguen una distribución normal / exponencial / logarítmica / no presenta distribución conocida debido a que (usar información sobre el descripción de cada variable)....
        - La mayoría de los ejemplos de las variables I, J y K tienen valor 0.
        - La variable M presenta gran cantidad de outliers.
        - Se observó que las variables categóricas E y F están bien balanceadas / presentan un gran desbalance.
        - Del gráfico de correlaciones se puede inferir que X, Y y Z tienen una correlación alta/baja entre ellas.



### 4. Preparación de Datos [1 Punto]

*Esta sección consiste en generar los distintos pasos para preparar sus datos con el fin de luego poder crear su modelo.*

Dentro de los aspectos minimos que se les solicitarán en este desarrollo estan:

- Definir preprocesadores para datos categóricos y ordinales.
- Setear las transformaciones en un `ColumnTransformer`.
- Transformar todo el datset.

**Notas**: 

- Las variables `id` y `fraud_bool` no deben ser transformadas, pero si pasadas a la siguiente etapa. (Estudiar la transformación `passthrough`).
- La salida de la transformación debe ser un pandas dataframe. Para esto, investiguen acerca de la nueva [`set_output` API](https://scikit-learn.org/stable/auto_examples/miscellaneous/plot_set_output.html). Si se levanta un error por matriz sparse, desactiven su uso en la transformación correspondiente.

### 5. Visualización en baja dimensionalidad [1 Punto]

Programar un script para proyectar los datos en baja dimensionalidad usando `PCA` o `UMAP`.

**Notas**: 

- Para realizar las proyecciones no consideren la variable `fraud_bool`, ya que esto provocaría un _data leakage_ (i.e., usar información del futuro para analizar el mismo fenómeno que quieren descubrir).
- Utilice un muestreo significativo del dataframe (10k a 25k ejemplos) para generar las proyecciones. De lo contrario, probablemente el proyector no sea capaz de asignar todos los recursos necesarios y lance alguna excepción.

Luego, por cada segmento etario generado en la sección 2, implemente un gráfico de dispersión (idealmente en plotly express) la proyección en 2D generada y coloree cada punto según la etiqueta `fraud_bool`.

Reportar si existen patrones y/o relaciones interesantes con respecto a variables de interés.


### 6. Explorar Anomalías [1.5 Puntos]

Proponer una técnica para detectar clientes fraudulentos en base a detección de anomalías. 
Al igual que el punto anterior, la técnica que proponga debe ser aplicada por separado en los diferentes segmentos etarios descritos en el punto 2 (es decir, debe entrenar un detector de anomalías para cada uno de estos grupos). 

Luego, al igual que el punto anterior, genere un gráfico de dispersión con las proyecciones 2D en donde el coloreado sea ahora las etiquetas predichas por el detector de anomalías.

Por último, usando queries de numpy/pandas, calcule dos ratios (P y R):

- P: Cantidad de ejemplos predichos correctamente como fraude / cantidad total de datos predichos como fraude.
- R: Cantidad de ejemplos predichos correctamente como fraude / cantidad total de ejemplos que eran realmente fraude.


De los resultados obtenidos responda las siguientes preguntas:

- ¿Qué significan los ratios y sus valores?
- ¿Qué tan correctas fueron las predicciones realizadas por su modelo según los ratios calculados?
- ¿Son coherentes los resultados obtenidos?.
- ¿Cómo se comportan los casos de fraudes para los diferentes rangos etarios?

Justifique cada una de sus respuestas.

---

## Esquema de la Tarea


Pueden usar el siguiente esquema para organizar la tarea (y borrar todo lo anterior).
Obviamente **no deben limitarse a lo que está escrito en esta**: puede incrementar en caso de más técnicas y obviar algunas partes en caso que alguna y otro punto no aplique a su problema.

Pueden borrar las instrucciones anteriores y quedarse solo con lo que viene a continuación.

---
<br>

<br>

<br>



# Proyecto

### Equipo:

- Diego Dominguez
- Arturo Lazcano


### Link de repositorio de GitHub: https://github.com/DDominguezD/MDS7202




## 1. Introducción


En este proyecto se trabajarán temas relacionados a la ciencia de datos en un dataset sobre fraude bancario proveniente de la conferencia anual NeurIPS 2022.  
Como es usual, el desarrollo del trabajo se dividirá en áreas como lectura y concatenaciones de dataframes a utilizar, realizar un análisis exploratorio, preprocesar los datos, probar técnicas de reducción de dimensionalidad para visualizar datos y, por último, probar una variedad de técnicas de detección de anomalías y clasificadores con el fin de crear un algoritmo que pueda detectar un fraude de mejor manera.  

Los datasets a trabajar proveen información sobre los ingresos de los clientes, edad, tipos de pago, tiempo con la sesión del banco abierta, entre otros. Así, de las 35 columnas presente, son 5 tipo 'Object' de tipo categóricas y 29 columnas numéricas.  
Se pueden apreciar varias características de los clientes en estos datos, donde la variable más importante, pues es la que se quiere predecir, es la de si hubo fraude o no. A pesar de esto, es posible esperar un gran desbalance en los datos pues la mayoría de clientes no realizan acciones fraudulentas, por lo cual, se piensa que el desbalance de los datos será uno de los mayores probelmas de este proyecto.  

Con tal de cumplir el objetivo de predecir fraudes, se hará un extenso manejo de los datos y uso de algoritmos de machine learning. Todo esto en Python junto con sus librerías clásicas de ciencia de datos.

## 2. Lectura y Manejo de Datos




En esta sección, se empieza importando las librerías que serán usadas a lo largo de este proyecto. Estan enfocadas tanto a cálculos numéricos, visualización y manejo de dataframes, gráficos y machine learning.
Luego, se realiza la lectura de los 3 dataframes cargándolos desde un Drive. Así, se procede a explorar las dimensiones, columnas y tipos de datos que poseen estos dataframes con el objetivo de concatenarlos en uno solo.

Finalmente, después de la concatenación, se revisan datos nulos y duplicados con el fin de limpiar el dataframe y decidir qué hacer con los duplicados. Por último, se genera una nueva columna llamada `segmentación_etaria`, y separa a los clientes del banco en rangos de joven (Menor a 18 años), adulto-joven (18 - 26 años), adulto (27- 59 años) y personamayor (60 años o más).

In [1]:
# Librerías principales
#!pip install --upgrade pandas
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches
import random
from IPython.display import display, Markdown, Latex

# Machine learning (sklearn)
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, OneHotEncoder, OrdinalEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.tree import DecisionTreeClassifier, plot_tree
from sklearn.ensemble import BaggingClassifier, RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, f1_score, classification_report, roc_curve, confusion_matrix
from sklearn import svm

# Gráficos
#!pip install --upgrade plotly
import plotly.express as px
from plotly.subplots import make_subplots

# Alertas
import warnings
#warnings.filterwarnings('ignore')

# EDA
!pip install ydata-profiling
from ydata_profiling import ProfileReport

# UMAP
!pip install umap-learn
import umap

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


### 2.1 Cargar datos en dataframes

In [2]:
# Si usted está utilizando Colabolatory le puede ser útil este código para cargar los archivos.
try:
    from google.colab import drive
    drive.mount("/content/drive")
    path = 'Dirección donde tiene los archivos en el Drive'
except: 
    print('Ignorando conexión drive-colab')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [3]:
# Lectura de dataframes por separado
df_1 = pd.read_parquet('/content/drive/MyDrive/df_1.parquet')
df_2 = pd.read_parquet('/content/drive/MyDrive/df_2.parquet')
df_mail_phone=pd.read_parquet('/content/drive/MyDrive/df_email_phone.parquet')

Los dataframes df_1 y df_2 contienen información de múltiples clientes, mientras que el dataframe df_mail_phone entrega características exclusivas relacionadas a los emails y teléfonos de algunos clientes. Esto puede provocar que haya filas de las que no se dispone esta información, además los primeros dos dataframes contienen información no necesariamente disjunta, así que puede se presente repeticiones de clientes.

Para asegurarnos de esto, a continuación se realizará un análisis exploratorio exhustivo de las bases de datos disponibles.

### 2.2 Exploración inicial

In [4]:
# Dimensiones y vista inicial del primer dataframe
filas, columnas = df_1.shape
print(f'El dataframe df_1 tiene {filas} filas y {columnas} columnas.')

df_1.head(5)

El dataframe df_1 tiene 389782 filas y 29 columnas.


id  fraud_bool  income  prev_address_months_count  \
104446  729517           0     0.7                         -1   
269483  149585           0     0.8                         -1   
4102     64486           0     0.9                         -1   
351767  825283           0     0.5                         -1   
126377    8308           0     0.9                         -1   

        current_address_months_count  customer_age  days_since_request  \
104446                           305            60            0.030059   
269483                           140            50            0.015659   
4102                             171            50            0.001409   
351767                            85            30            0.027292   
126377                            39            30            0.010945   

        intended_balcon_amount payment_type  zip_count_4w  ...  \
104446               -1.599455           AC           990  ...   
269483                3.951994           AA          1269  ...   
4102                 28.159779           AB          4430  ...   
351767               -1.310498           AB          1698  ...   
126377               -1.450972           AC           569  ...   

        proposed_credit_limit  foreign_request    source  \
104446                  500.0                0  INTERNET   
269483                  200.0                0  INTERNET   
4102                   1500.0                0  INTERNET   
351767                  200.0                0  INTERNET   
126377                  200.0                0  INTERNET   

        session_length_in_minutes device_os  keep_alive_session  \
104446                   8.865992   windows                   0   
269483                   4.654872     linux                   1   
4102                     3.720953     linux                   1   
351767                   2.912670   windows                   1   
126377                   2.283680     other                   1   

       device_fraud_count  month        x1        x2  
104446                  0      2 -0.245425  0.568811  
269483                  0      7  0.009336 -2.096682  
4102                    0      1  2.229616 -0.005823  
351767                  0      3 -0.193945  0.861207  
126377                  0      3 -2.440650  0.354986  

[5 rows x 29 columns]

In [5]:
# Información del primer dataframe
df_1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 389782 entries, 104446 to 193209
Data columns (total 29 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   id                            389782 non-null  int64  
 1   fraud_bool                    389782 non-null  int64  
 2   income                        389782 non-null  float64
 3   prev_address_months_count     389782 non-null  int64  
 4   current_address_months_count  389782 non-null  int64  
 5   customer_age                  389782 non-null  int64  
 6   days_since_request            389782 non-null  float64
 7   intended_balcon_amount        389782 non-null  float64
 8   payment_type                  389782 non-null  object 
 9   zip_count_4w                  389782 non-null  int64  
 10  velocity_6h                   389782 non-null  float64
 11  velocity_24h                  389782 non-null  float64
 12  velocity_4w                   389782 no

In [6]:
# Dimensiones y vista inicial del segundo dataframe
filas, columnas = df_2.shape
print(f'El dataframe df_2 tiene {filas} filas y {columnas} columnas.')

df_2.head(5)

El dataframe df_2 tiene 747410 filas y 29 columnas.


id  fraud_bool  income  prev_address_months_count  \
699954  303612           0     0.8                         -1   
701206  568066           0     0.9                         -1   
270304  893344           0     0.7                         -1   
346538  247688           0     0.8                         -1   
147297  674478           0     0.8                         -1   

        current_address_months_count  customer_age  days_since_request  \
699954                            21            40            0.010837   
701206                            70            50            0.711485   
270304                           133            50            0.006856   
346538                           207            50            0.026194   
147297                            24            50            0.012646   

        intended_balcon_amount payment_type  zip_count_4w  ...  \
699954               -0.853976           AD           239  ...   
701206               25.137456           AA          1873  ...   
270304               -0.815330           AC           675  ...   
346538               -0.736615           AB           999  ...   
147297               10.744253           AA          3198  ...   

        proposed_credit_limit  foreign_request    source  \
699954                  200.0                0  INTERNET   
701206                  200.0                0  INTERNET   
270304                 1000.0                0  INTERNET   
346538                  200.0                0  INTERNET   
147297                  500.0                1  INTERNET   

        session_length_in_minutes device_os  keep_alive_session  \
699954                  48.127390     other                   0   
701206                   8.792273   windows                   1   
270304                  15.916225     other                   1   
346538                   0.509942     linux                   1   
147297                  10.562945     other                   0   

       device_fraud_count  month        x1        x2  
699954                  0      3  0.236931  0.077061  
701206                  0      3  0.571536  1.151563  
270304                  0      7  0.224252 -0.590459  
346538                  0      6  0.648528 -0.315911  
147297                  0      6  0.041475  0.383332  

[5 rows x 29 columns]

In [7]:
# Información del segundo dataframe
df_2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 747410 entries, 699954 to 384611
Data columns (total 29 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   id                            747410 non-null  int64  
 1   fraud_bool                    747410 non-null  int64  
 2   income                        747410 non-null  float64
 3   prev_address_months_count     747410 non-null  int64  
 4   current_address_months_count  747410 non-null  int64  
 5   customer_age                  747410 non-null  int64  
 6   days_since_request            747410 non-null  float64
 7   intended_balcon_amount        747410 non-null  float64
 8   payment_type                  747410 non-null  object 
 9   zip_count_4w                  747410 non-null  int64  
 10  velocity_6h                   747410 non-null  float64
 11  velocity_24h                  747410 non-null  float64
 12  velocity_4w                   747410 no

In [8]:
# Dimensiones y vista inicial del tercer dataframe
filas, columnas = df_mail_phone.shape
print(f'El dataframe df_mail_phone tiene {filas} filas y {columnas} columnas.')

df_mail_phone.head(5)

El dataframe df_mail_phone tiene 934730 filas y 7 columnas.


id  name_email_similarity  date_of_birth_distinct_emails_4w  \
987231  624588               0.555653                                15   
79954   620810               0.849718                                 2   
567130  580633               0.110898                                 3   
500891    6379               0.671250                                15   
55399   366511               0.772932                                 7   

        email_is_free  device_distinct_emails_8w  phone_home_valid  \
987231              0                          1                 0   
79954               0                          1                 0   
567130              0                          1                 1   
500891              0                          1                 0   
55399               0                          1                 0   

        phone_mobile_valid  
987231                   1  
79954                    0  
567130                   0  
500891                   1  
55399                    1

In [9]:
# Información del tercer dataframe
df_mail_phone.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 934730 entries, 987231 to 986114
Data columns (total 7 columns):
 #   Column                            Non-Null Count   Dtype  
---  ------                            --------------   -----  
 0   id                                934730 non-null  int64  
 1   name_email_similarity             934730 non-null  float64
 2   date_of_birth_distinct_emails_4w  934730 non-null  int64  
 3   email_is_free                     934730 non-null  int64  
 4   device_distinct_emails_8w         934730 non-null  int64  
 5   phone_home_valid                  934730 non-null  int64  
 6   phone_mobile_valid                934730 non-null  int64  
dtypes: float64(1), int64(6)
memory usage: 57.1 MB


### 2.3 Unir dataframes

In [10]:
# Unión primeros dos dataframes
df_ejemplos=pd.merge(df_1, df_2, how='outer')#, left_on='id', right_on='id')
df_ejemplos

id  fraud_bool  income  prev_address_months_count  \
0        729517           0     0.7                         -1   
1        149585           0     0.8                         -1   
2         64486           0     0.9                         -1   
3        825283           0     0.5                         -1   
4          8308           0     0.9                         -1   
...         ...         ...     ...                        ...   
1098145  707063           0     0.1                         -1   
1098146  784606           0     0.5                         -1   
1098147  216375           0     0.8                         46   
1098148   41856           0     0.7                         25   
1098149  988191           0     0.9                         -1   

         current_address_months_count  customer_age  days_since_request  \
0                                 305            60            0.030059   
1                                 140            50            0.015659   
2                                 171            50            0.001409   
3                                  85            30            0.027292   
4                                  39            30            0.010945   
...                               ...           ...                 ...   
1098145                           260            60            0.026023   
1098146                            30            50            0.052916   
1098147                             0            30            0.005857   
1098148                            18            30            1.426644   
1098149                            56            30            0.006165   

         intended_balcon_amount payment_type  zip_count_4w  ...  \
0                     -1.599455           AC           990  ...   
1                      3.951994           AA          1269  ...   
2                     28.159779           AB          4430  ...   
3                     -1.310498           AB          1698  ...   
4                     -1.450972           AC           569  ...   
...                         ...          ...           ...  ...   
1098145               -0.587295           AB          1795  ...   
1098146               -1.249585           AB          1039  ...   
1098147               -0.380745           AB           918  ...   
1098148               10.441915           AA          1188  ...   
1098149               -1.103130           AD          1922  ...   

         proposed_credit_limit  foreign_request    source  \
0                        500.0                0  INTERNET   
1                        200.0                0  INTERNET   
2                       1500.0                0  INTERNET   
3                        200.0                0  INTERNET   
4                        200.0                0  INTERNET   
...                        ...              ...       ...   
1098145                 1000.0                0  INTERNET   
1098146                  200.0                0  INTERNET   
1098147                 1500.0                0  INTERNET   
1098148                  200.0                0  INTERNET   
1098149                  200.0                0  INTERNET   

         session_length_in_minutes  device_os  keep_alive_session  \
0                         8.865992    windows                   0   
1                         4.654872      linux                   1   
2                         3.720953      linux                   1   
3                         2.912670    windows                   1   
4                         2.283680      other                   1   
...                            ...        ...                 ...   
1098145                   5.780150    windows                   0   
1098146                   3.359270  macintosh                   1   
1098147                   4.821661    windows                   0   
1098148                   2.899507      other                   1   
1098149               

### 2.4 Unir nuevas variables

In [11]:
# Unión final de datos
df = pd.merge(df_ejemplos, df_mail_phone, how='outer', left_on='id', right_on='id')

# Visualizar dataframe final
filas, columnas = df.shape
print(f'El dataframe df tiene {filas} filas y {columnas} columnas.')

df.head()

El dataframe df tiene 1159109 filas y 35 columnas.


id  fraud_bool  income  prev_address_months_count  \
0  729517         0.0     0.7                       -1.0   
1  149585         0.0     0.8                       -1.0   
2   64486         0.0     0.9                       -1.0   
3  825283         0.0     0.5                       -1.0   
4    8308         0.0     0.9                       -1.0   

   current_address_months_count  customer_age  days_since_request  \
0                         305.0          60.0            0.030059   
1                         140.0          50.0            0.015659   
2                         171.0          50.0            0.001409   
3                          85.0          30.0            0.027292   
4                          39.0          30.0            0.010945   

   intended_balcon_amount payment_type  zip_count_4w  ...  device_fraud_count  \
0               -1.599455           AC         990.0  ...                 0.0   
1                3.951994           AA        1269.0  ...                 0.0   
2               28.159779           AB        4430.0  ...                 0.0   
3               -1.310498           AB        1698.0  ...                 0.0   
4               -1.450972           AC         569.0  ...                 0.0   

   month        x1        x2 name_email_similarity  \
0    2.0 -0.245425  0.568811              0.883485   
1    7.0  0.009336 -2.096682              0.113208   
2    1.0  2.229616 -0.005823              0.792797   
3    3.0 -0.193945  0.861207              0.865082   
4    3.0 -2.440650  0.354986              0.450024   

   date_of_birth_distinct_emails_4w email_is_free  device_distinct_emails_8w  \
0                               2.0           1.0                        1.0   
1                               4.0           1.0                        1.0   
2                               4.0           0.0                        1.0   
3                              10.0           0.0                        1.0   
4                              13.0           1.0                        1.0   

   phone_home_valid  phone_mobile_valid  
0               0.0                 1.0  
1               1.0                 1.0  
2               0.0                 1.0  
3               0.0                 1.0  
4               0.0                 0.0  

[5 rows x 35 columns]

In [12]:
# Información del dataframe final
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1159109 entries, 0 to 1159108
Data columns (total 35 columns):
 #   Column                            Non-Null Count    Dtype  
---  ------                            --------------    -----  
 0   id                                1159109 non-null  int64  
 1   fraud_bool                        1098150 non-null  float64
 2   income                            1098150 non-null  float64
 3   prev_address_months_count         1098150 non-null  float64
 4   current_address_months_count      1098150 non-null  float64
 5   customer_age                      1098150 non-null  float64
 6   days_since_request                1098150 non-null  float64
 7   intended_balcon_amount            1098150 non-null  float64
 8   payment_type                      1098150 non-null  object 
 9   zip_count_4w                      1098150 non-null  float64
 10  velocity_6h                       1098150 non-null  float64
 11  velocity_24h                      109

In [13]:
# Descripción del dataframe final
df.describe()

id    fraud_bool        income  prev_address_months_count  \
count  1.159109e+06  1.098150e+06  1.098150e+06               1.098150e+06   
mean   5.001192e+05  1.098666e-02  5.784957e-01               1.464209e+01   
std    2.887479e+05  1.042399e-01  2.882668e-01               4.296932e+01   
min    0.000000e+00  0.000000e+00  1.000000e-01              -1.000000e+00   
25%    2.499580e+05  0.000000e+00  3.000000e-01              -1.000000e+00   
50%    5.003040e+05  0.000000e+00  6.000000e-01              -1.000000e+00   
75%    7.502040e+05  0.000000e+00  8.000000e-01              -1.000000e+00   
max    9.999990e+05  1.000000e+00  9.000000e-01               3.990000e+02   

       current_address_months_count  customer_age  days_since_request  \
count                  1.098150e+06  1.098150e+06        1.098150e+06   
mean                   9.918096e+01  4.134635e+01        9.018398e-01   
std                    9.406535e+01  1.376732e+01        5.002738e+00   
min                   -1.000000e+00  1.000000e+01        3.112791e-08   
25%                    2.700000e+01  3.000000e+01        7.430442e-03   
50%                    6.400000e+01  5.000000e+01        1.565511e-02   
75%                    1.540000e+02  5.000000e+01        2.688838e-02   
max                    4.290000e+02  9.000000e+01        7.657750e+01   

       intended_balcon_amount  zip_count_4w   velocity_6h  ...  \
count            1.098150e+06  1.098150e+06  1.098150e+06  ...   
mean             8.565908e+00  1.518424e+03  5.491743e+03  ...   
std              2.054504e+01  9.661186e+02  2.940254e+03  ...   
min             -1.553733e+01  1.000000e+00 -1.741097e+02  ...   
25%             -1.179867e+00  8.860000e+02  3.336792e+03  ...   
50%             -8.340512e-01  1.209000e+03  5.192766e+03  ...   
75%             -6.265769e-02  1.845000e+03  7.368170e+03  ...   
max              1.127025e+02  6.650000e+03  1.675496e+04  ...   

       device_fraud_count         month            x1            x2  \
count           1098150.0  1.098150e+06  1.098150e+06  1.098150e+06   
mean                  0.0  3.656341e+00  1.497024e-02  1.119093e-02   
std                   0.0  2.117036e+00  1.013277e+00  1.012556e+00   
min                   0.0  0.000000e+00 -4.977864e+00 -4.846414e+00   
25%                   0.0  2.000000e+00 -6.668884e-01 -6.700486e-01   
50%                   0.0  4.000000e+00  8.903174e-03  5.188557e-03   
75%                   0.0  5.000000e+00  6.869599e-01  6.836585e-01   
max                   0.0  7.000000e+00  6.434867e+00  6.542492e+00   

       name_email_similarity  date_of_birth_distinct_emails_4w  email_is_free  \
count           1.079382e+06                      1.079382e+06   1.079382e+06   
mean            4.875715e-01                      7.777006e+00   5.181715e-01   
std             2.914606e-01                      4.815093e+00   4.996699e-01   
min             5.024707e-08                      0.000000e+00   0.000000e+00   
25%             2.144731e-01                      4.000000e+00   0.000000e+00   
50%             4.858862e-01                      7.000000e+00   1.000000e+00   
75%             7.544564e-01                      1.100000e+01   1.000000e+00   
max             1.000000e+00                      3.900000e+01   1.000000e+00   

       device_distinct_emails_8w  phone_home_valid  phone_mobile_valid  
count               1.079382e+06      1.079382e+06        1.079382e+06  
mean                1.022178e+00      4.934722e-01        8.570163e-01  
std                 1.914559e-01      4.999576e-01        3.500564e-01  
min                -1.000000e+00      0.000000e+00        0.000000e+00  
25%                 1.000000e+00      0.000000e+00        1.000000e+00  
50%                 1.000000e+00      0.000000e+00        1.000000e+00  
75%                 1.000000e+00      1.000000e+00        1.000000e+00  
max                 2.000000e+00      1.000000e+00        1.000000e+00  

[8 rows x 30 columns]

### 2.5 Verificar nulos y duplicados, generar explicaciones sobre sus fuentes y proponer soluciones.



In [14]:
df_ejemplos.isna().sum()

id                              0
fraud_bool                      0
income                          0
prev_address_months_count       0
current_address_months_count    0
customer_age                    0
days_since_request              0
intended_balcon_amount          0
payment_type                    0
zip_count_4w                    0
velocity_6h                     0
velocity_24h                    0
velocity_4w                     0
bank_branch_count_8w            0
employment_status               0
credit_risk_score               0
housing_status                  0
bank_months_count               0
has_other_cards                 0
proposed_credit_limit           0
foreign_request                 0
source                          0
session_length_in_minutes       0
device_os                       0
keep_alive_session              0
device_fraud_count              0
month                           0
x1                              0
x2                              0
dtype: int64

In [15]:
df_mail_phone.isna().sum()

id                                  0
name_email_similarity               0
date_of_birth_distinct_emails_4w    0
email_is_free                       0
device_distinct_emails_8w           0
phone_home_valid                    0
phone_mobile_valid                  0
dtype: int64

In [16]:
# Suma cantidad de valores nulos por columna
df.isna().sum()

id                                      0
fraud_bool                          60959
income                              60959
prev_address_months_count           60959
current_address_months_count        60959
customer_age                        60959
days_since_request                  60959
intended_balcon_amount              60959
payment_type                        60959
zip_count_4w                        60959
velocity_6h                         60959
velocity_24h                        60959
velocity_4w                         60959
bank_branch_count_8w                60959
employment_status                   60959
credit_risk_score                   60959
housing_status                      60959
bank_months_count                   60959
has_other_cards                     60959
proposed_credit_limit               60959
foreign_request                     60959
source                              60959
session_length_in_minutes           60959
device_os                         

El dataframe `df_ejemplos` corresponde a la unión entre `df_1` y `df_2`. Como se puede observar en la celda de más arriba, este no posee valores nulos, al igual que `df_mail_phone`. No obstante, la combinación de ambos, es decir, el dataframe `df`, sí presenta entre 60 mil y 80 mil valores nulos en todas las columnas excepto en el campo `id`. Esto se debe a que `df_ejemplos` y `df_mail_phone` no comparten todos los `id`, por lo tanto, al combinarlos, habrá personas con datos faltantes. Para solucionar esto, se eliminará todos las filas de clientes que posean algún dato nulo.

In [17]:
# Revisión de id's duplicados
df_duplicados=df[df['id'].duplicated()]

# Columna size indica cantidad de veces repetido el id
df_ejemplos_duplicados=df_duplicados.groupby(['id'], as_index=False).size()
df_ejemplos_duplicados

id  size
0          19     2
1          37     2
2          39     3
3          55     2
4          73     3
...       ...   ...
58274  999857     3
58275  999870     3
58276  999901     3
58277  999962     3
58278  999996     3

[58279 rows x 2 columns]

Además, como se mencionó anteriormente, los dataframes `df_1` y `df_2` pueden compartir datos, lo que resulta en la presencia de múltiples registros duplicados de clientes en el dataframe combinado. En la tabla de arriba se muestra el ID de cada persona junto con la cantidad de veces que aparece en el dataframe `df`.
Para abordar esta situación, se conservará únicamente la primera fila de cada cliente y eliminando las filas repetidas.

### 2.6 Limpiar

In [18]:
# Limpieza de datos dulpicados (manteniendo el primero)
df.drop_duplicates(subset=['id'], keep='first', inplace=True)
df.sort_values(by='id').head()

id  fraud_bool  income  prev_address_months_count  \
235230    0         0.0     0.2                      326.0   
1102276   1         NaN     NaN                        NaN   
67648     2         0.0     0.9                       -1.0   
764072    3         0.0     0.8                       23.0   
926894    4         0.0     0.1                       -1.0   

         current_address_months_count  customer_age  days_since_request  \
235230                           20.0          60.0            0.022212   
1102276                           NaN           NaN                 NaN   
67648                            41.0          50.0            0.002588   
764072                            9.0          50.0            0.004026   
926894                           30.0          30.0            0.011687   

         intended_balcon_amount payment_type  zip_count_4w  ...  \
235230                -0.758101           AB         647.0  ...   
1102276                     NaN          NaN           NaN  ...   
67648                 17.710636           AA         815.0  ...   
764072                18.224113           AA        1204.0  ...   
926894                -1.434154           AB        1182.0  ...   

         device_fraud_count  month        x1        x2 name_email_similarity  \
235230                  0.0    3.0 -0.282223 -0.137122              0.369510   
1102276                 NaN    NaN       NaN       NaN              0.858929   
67648                   0.0    4.0 -0.029179  1.089166              0.166457   
764072                  0.0    2.0  2.178365  1.301696              0.250290   
926894                  0.0    7.0  1.472099  0.367675              0.619087   

         date_of_birth_distinct_emails_4w email_is_free  \
235230                                2.0           0.0   
1102276                               6.0           0.0   
67648                                 7.0           1.0   
764072                                1.0           0.0   
926894                                8.0           0.0   

         device_distinct_emails_8w  phone_home_valid  phone_mobile_valid  
235230                         2.0               0.0                 0.0  
1102276                        1.0               1.0                 1.0  
67648                          1.0               1.0                 1.0  
764072                         1.0               0.0                 1.0  
926894                         1.0               0.0                 1.0  

[5 rows x 35 columns]

In [19]:
# Dropear valores nulos
df.dropna(inplace=True)
df.tail()

id  fraud_bool  income  prev_address_months_count  \
1098145  707063         0.0     0.1                       -1.0   
1098146  784606         0.0     0.5                       -1.0   
1098147  216375         0.0     0.8                       46.0   
1098148   41856         0.0     0.7                       25.0   
1098149  988191         0.0     0.9                       -1.0   

         current_address_months_count  customer_age  days_since_request  \
1098145                         260.0          60.0            0.026023   
1098146                          30.0          50.0            0.052916   
1098147                           0.0          30.0            0.005857   
1098148                          18.0          30.0            1.426644   
1098149                          56.0          30.0            0.006165   

         intended_balcon_amount payment_type  zip_count_4w  ...  \
1098145               -0.587295           AB        1795.0  ...   
1098146               -1.249585           AB        1039.0  ...   
1098147               -0.380745           AB         918.0  ...   
1098148               10.441915           AA        1188.0  ...   
1098149               -1.103130           AD        1922.0  ...   

         device_fraud_count  month        x1        x2 name_email_similarity  \
1098145                 0.0    5.0  0.397897 -0.103541              0.870118   
1098146                 0.0    1.0  1.713050 -1.911669              0.094715   
1098147                 0.0    3.0 -0.727536 -1.540993              0.437026   
1098148                 0.0    3.0  1.642429 -0.026783              0.449010   
1098149                 0.0    4.0 -1.662247  1.576214              0.769959   

         date_of_birth_distinct_emails_4w email_is_free  \
1098145                               3.0           1.0   
1098146                               5.0           0.0   
1098147                              13.0           0.0   
1098148                              15.0           0.0   
1098149                              10.0           1.0   

         device_distinct_emails_8w  phone_home_valid  phone_mobile_valid  
1098145                        1.0               0.0                 0.0  
1098146                        1.0               1.0                 1.0  
1098147                        1.0               0.0                 1.0  
1098148                        1.0               1.0                 1.0  
1098149                        1.0               1.0                 1.0  

[5 rows x 35 columns]

In [20]:
# Dimensiones y vista inicial del dataframe final
filas, columnas = df.shape
print(f'El dataframe df tiene {filas} filas y {columnas} columnas.')
df.describe()

El dataframe df tiene 873771 filas y 35 columnas.


id     fraud_bool         income  prev_address_months_count  \
count  873771.000000  873771.000000  873771.000000              873771.000000   
mean   499986.082237       0.011012       0.578727                  14.664555   
std    288603.604762       0.104359       0.288292                  43.026325   
min         0.000000       0.000000       0.100000                  -1.000000   
25%    249948.500000       0.000000       0.300000                  -1.000000   
50%    500153.000000       0.000000       0.600000                  -1.000000   
75%    749761.500000       0.000000       0.800000                  -1.000000   
max    999999.000000       1.000000       0.900000                 399.000000   

       current_address_months_count   customer_age  days_since_request  \
count                 873771.000000  873771.000000        8.737710e+05   
mean                      99.234073      41.344013        9.026673e-01   
std                       94.047441      13.762575        4.996813e+00   
min                       -1.000000      10.000000        3.112791e-08   
25%                       27.000000      30.000000        7.448993e-03   
50%                       64.000000      50.000000        1.567109e-02   
75%                      154.000000      50.000000        2.693016e-02   
max                      429.000000      90.000000        7.657750e+01   

       intended_balcon_amount   zip_count_4w    velocity_6h  ...  \
count           873771.000000  873771.000000  873771.000000  ...   
mean                 8.554425    1517.752437    5490.980468  ...   
std                 20.528247     965.194858    2940.464800  ...   
min                -15.537329       1.000000    -174.109691  ...   
25%                 -1.179763     886.000000    3335.150770  ...   
50%                 -0.834561    1208.000000    5190.987335  ...   
75%                 -0.065957    1844.000000    7369.109285  ...   
max                112.702504    6650.000000   16754.959024  ...   

       device_fraud_count          month             x1             x2  \
count            873771.0  873771.000000  873771.000000  873771.000000   
mean                  0.0       3.657684       0.014022       0.011191   
std                   0.0       2.117050       1.013359       1.012448   
min                   0.0       0.000000      -4.977864      -4.846414   
25%                   0.0       2.000000      -0.668335      -0.670306   
50%                   0.0       4.000000       0.006772       0.005482   
75%                   0.0       5.000000       0.686693       0.683431   
max                   0.0       7.000000       6.434867       6.542492   

       name_email_similarity  date_of_birth_distinct_emails_4w  email_is_free  \
count           8.737710e+05                     873771.000000  873771.000000   
mean            4.877150e-01                          7.773291       0.518628   
std             2.914227e-01                          4.815301       0.499653   
min             5.024707e-08                          0.000000       0.000000   
25%             2.146826e-01                          4.000000       0.000000   
50%             4.860334e-01                          7.000000       1.000000   
75%             7.546222e-01                         11.000000       1.000000   
max             1.000000e+00                         39.000000       1.000000   

       device_distinct_emails_8w  phone_home_valid  phone_mobile_valid  
count              873771.000000     873771.000000       873771.000000  
mean                    1.022184          0.493474            0.856930  
std                     0.191700          0.499958            0.350144  
min                    -1.000000          0.000000            0.000000  
25%                     1.000000          0.000000            1.000000  
50%                     1.000000          0.000000            1.000000  
75%                     1.000000          1.000000            1.000000  
max                     2.0

Una vez realizada las modificaciones anteriores, el dataframe limpio tiene 873.771 filas, es decir, se eliminaron 285.338 ejemplos. 

### 2.7 Segmentación etaria

In [21]:
# Definir rango de edades
bins= [0,18,26,60,100]
# Definir nombre a los rangos
labels = ['Joven', 'Adulto-Joven', 'Adulto', 'Persona Mayor']
# Añadir columna al dataframe
df['segmentacion_etaria'] = pd.cut(df['customer_age'], bins=bins, labels=labels, right=False)
df.head()

id  fraud_bool  income  prev_address_months_count  \
0  729517         0.0     0.7                       -1.0   
1  149585         0.0     0.8                       -1.0   
2   64486         0.0     0.9                       -1.0   
3  825283         0.0     0.5                       -1.0   
4    8308         0.0     0.9                       -1.0   

   current_address_months_count  customer_age  days_since_request  \
0                         305.0          60.0            0.030059   
1                         140.0          50.0            0.015659   
2                         171.0          50.0            0.001409   
3                          85.0          30.0            0.027292   
4                          39.0          30.0            0.010945   

   intended_balcon_amount payment_type  zip_count_4w  ...  month        x1  \
0               -1.599455           AC         990.0  ...    2.0 -0.245425   
1                3.951994           AA        1269.0  ...    7.0  0.009336   
2               28.159779           AB        4430.0  ...    1.0  2.229616   
3               -1.310498           AB        1698.0  ...    3.0 -0.193945   
4               -1.450972           AC         569.0  ...    3.0 -2.440650   

         x2  name_email_similarity date_of_birth_distinct_emails_4w  \
0  0.568811               0.883485                              2.0   
1 -2.096682               0.113208                              4.0   
2 -0.005823               0.792797                              4.0   
3  0.861207               0.865082                             10.0   
4  0.354986               0.450024                             13.0   

   email_is_free device_distinct_emails_8w  phone_home_valid  \
0            1.0                       1.0               0.0   
1            1.0                       1.0               1.0   
2            0.0                       1.0               0.0   
3            0.0                       1.0               0.0   
4            1.0                       1.0               0.0   

   phone_mobile_valid  segmentacion_etaria  
0                 1.0        Persona Mayor  
1                 1.0               Adulto  
2                 1.0               Adulto  
3                 1.0               Adulto  
4                 0.0               Adulto  

[5 rows x 36 columns]

## 3. Análisis Exploratorio de Datos

### 3.1 Análisis del EDA

En esta sección se raliza el análisis exploratorio de datos (EDA). Para ello, se hace uso de la librería ydata_profiling que provee un reporte automático transformado a un archivo html. Este código genera análisis sobre los tipos de datos, alertas (tipos de datos nulos, skewnees, entre otros), análisis univariado y multivariado y más.
Notar que este estudio del dataframe se verá reflejado en el archivo html y no en este archivo ipynb, sin embargo, se comentarán observaciones y posibles patrones encontrados en las columnas del dataset.

In [22]:
profile = ProfileReport(df, title="EDA Bodoque Bank")

#Esta celda esta comentada para que no se genere el HTML nuevamente
#profile.to_file("EDA_report.html")

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

A continuación, se mencionan las características más importantes que se obtuvieron del EDA:  


*   Se comprueba que no hay datos nulos ni **id's** duplicadas.
*   Las variables con mayor correlación tienen relación a datos temporales (o acciones realizadas hace cierto tiempo).
*   La columna `device_fraud_count` tiene un único valor (se puede eliminar del dataframe).
*   `customer_age` y `segmentacion_etaria` tienen una alta correlación positiva como es de esperar.
*   La columna `credit_risk_score` está altamente relacionada con `proposed_credit_limit`.
*   Las columnas `fraud_bool, foreign_request, source` y `device_distinct_emails_8w` poseen un gran desbalance en los datos únicos de cada una.
*   Las variables `id, x1` y `x2` tienen sólo valores únicos por cada cliente.
*   La variable `id` tiene una distribución uniforme entre $0$ y $999999$

Las columnas que no han sido nombradas en esta lista son aquellas que el algoritmo no ha detectado alguna alta correlación, patron, distribución ni alguna otra característica que sea importante destacar.  

Por otro lado, las columnas `x1` y `x2` parecen tener una distribución normal en el gráfico, sin embargo, esto se debería comprobar con el significado de estas columnas para analizar el sentido de esta hipótesis.  

Por último, se puede comprobar bajo la matriz de correlación que la variable target `fraud_bool` posee prácticamente correlación $0$ con las otras variabes, exceptuando `x1`y `x2` que posee un valor de $0.28$ aproximadamente. Esto quiere decir que, a priori, es díficil predecir un fraude analizando de manera simple alguna de estas columnas. 

## 4. Preprocesamiento

Para esta parte del trabajo, se genera un  `ColumnTransformer` que permita modificar los datos. Estas modificaciones son `OneHotEncoder`, `StandarScaler` y `OrdinalEncoder`.  

Para ver a cuales columnas aplicarles alguna transformación, se analizarán primero sus tipos y si tiene sentido transformar la variable. Es decir, la columna **"id"** y **"fraud_bool"** carecen de sentido para las transformaciones por lo que se mantendrán como vienen en el dataframe. Por otro lado, las columnas numéricas pasarán por un escaamiento, mientras que las de tipo **'Object'**, se verán si poseen pocos valores únicos (indicando que son características categóricas) para pasarlas a *one hot encoder*.  

Por último, se genera un `Pipeline` para guardar las transformaciones y así generar un nuevo dataframe procesado.

### 4.1 Declarar `ColumnTransformer`

In [23]:
# Verificar datos tipo 'object'
print('Valores únicos de las columnas de tipo object')
print(' ')
column_object=list(df.select_dtypes(['object']).columns)
for col in column_object:
    print(col+':', df[col].unique())

Valores únicos de las columnas de tipo object
 
payment_type: ['AC' 'AA' 'AB' 'AD' 'AE']
employment_status: ['CC' 'CF' 'CA' 'CB' 'CE' 'CD' 'CG']
housing_status: ['BB' 'BA' 'BE' 'BC' 'BD' 'BF' 'BG']
source: ['INTERNET' 'TELEAPP']
device_os: ['windows' 'linux' 'other' 'macintosh' 'x11']


In [24]:
# Columnas tipo float
column_float=list(df.select_dtypes(['float64']).columns)
column_float=column_float[1:] #No se considera la columna fraud_bool

In [25]:
# Preprocesamiento de datos
preprocessing_transformer = ColumnTransformer(transformers=[
    ('OneHotEncoder', OneHotEncoder(sparse_output=False), column_object),
    ('scaler', StandardScaler(), column_float),
    ('ordinal_encoder', OrdinalEncoder(), ['segmentacion_etaria'])], 
    remainder='passthrough')

In [26]:
# Crear pipeline
pipeline = Pipeline([('Preprocessing', preprocessing_transformer)])

### 4.2 Transformar datos

In [27]:
# Retornar datframe procesado
pipeline.set_output(transform="pandas")
df_preprocesado = pipeline.fit_transform(df)

## 5. Visualización en Baja Dimensionalidad

En esta sección, se espera poder visualizar los datos en baja dimensión ($\mathbb{R}^2$ en este caso). Para ello, se hará un muestreo aleatorio de datos (entre 10.000 y 25.000), para luego utilizar el algoritmo `UMAP`, que permite lograr nuestro objetivo.
Notar que se debe eliminar la variable `fraud_bool` pues este es un algoritmo no supervisado y no queremos entregarle información real o del futuro sobre el resultado de fraude.

Esta visualización de datos se hará con respecto a cada segmentación etaria para poder observar si existe algún patron entre esta variable y las demás para predecir, separar u observar algún patron sobre los fraudes.

### 5.1 Muestrear dataframe

In [28]:
# Seleccionar muestreo pequeño de datos al azar (replicable)
df_preprocesado_sample = df_preprocesado.sample(20000, random_state=2)
df_sample = df.sample(20000, random_state=2)

In [29]:
# Definir variable target y dropear id, target en el dataframe
labels = df_preprocesado_sample['remainder__fraud_bool']
df_preprocesado_sample.drop(['remainder__id', 'remainder__fraud_bool'], axis=1, inplace=True)

### 5.2 Proyectar y agregar proyecciones al dataframe de muestreo

In [30]:
# Crear umap
umap_ = umap.UMAP()
# Entrenar umap
features_umap_embedded = umap_.fit_transform(df_preprocesado_sample)
features_umap_embedded

array([[ 7.084492 ,  5.191807 ],
       [ 4.261039 ,  8.861938 ],
       [12.478038 , -3.7681532],
       ...,
       [ 7.9365864,  9.334736 ],
       [ 9.69015  ,  6.8153577],
       [ 7.8186436,  9.543944 ]], dtype=float32)

In [31]:
# Noten que se guarda la transformación en df y no en features.
df_sample["x_umap"] = features_umap_embedded[:, 0] 
df_sample["y_umap"] = features_umap_embedded[:, 1] 

In [32]:
# Pasar variable target a string por legenda del gráfico de plotly
df_sample['fraud_bool']=df_sample['fraud_bool'].astype(str)

### 5.3 Visualizar según rangos etarios

In [33]:
# Gráfico en dimensión reducida por segmentación entaria separado por fraude {0,1}
for edad in df_sample['segmentacion_etaria'].unique():
  fig = px.scatter(
    df_sample[df_sample['segmentacion_etaria']==edad],
    x="x_umap",
    y="y_umap",
    color='fraud_bool',
    hover_data=['credit_risk_score' , 'phone_mobile_valid']
  ) 
  # Cambiar dimensiones y título
  fig.update_layout(height=700, width=800, title_text="UMAP para "+edad)
  # Mostrar figura
  fig.show()

Para empezar, se obtuvo una muestra de 20.000 ejemplos aleatorios de la base de datos original. Una vez realizada la proyección del dataframe de muestreo usando UMAP, se obtuvo diferentes gráficos por segmentación etaria.
De estos gráficos, se puede visualizar que la cantidad de fraudes es muy infíma en relación a los datos totales. De hecho, para los clientes jóvenes no se registro ningún fraude. 

Aún así, para los rangos donde si hubo mayor casos de fraude como Adulto y y Persona mayor, se puede ver que los fraudes están repartidos uniformemente en los diferentes cumulos de puntos, por lo que realizar custering como DBSCAN para encontrar datos anomalos, no nos entregaría los datos correspondientes a los fraudes. 

## 6. Anomalías

### 6.1 Implementar detector de anomalías sobre dataframe de muestreo

In [34]:
# Definir características y variable target
X = df_preprocesado_sample
y = labels

# Observar balance de los datos
y.value_counts()

0.0    19782
1.0      218
Name: remainder__fraud_bool, dtype: int64

In [35]:
# Separacion datos de entrenamiento y testeo
X_train, X_test, y_train, y_test = train_test_split(X, y, shuffle=True, stratify=y, test_size=0.3, random_state=10) 

In [36]:
Models = [RandomForestClassifier(n_estimators=500, n_jobs=-1), LogisticRegression(), LogisticRegression(class_weight={0.0: 0.08, 1.0: 0.92})]
labels= ['Random Forest', 'Regresión Logística', 'Regresión Logística balanceada']
for i, clf in enumerate(Models):
  print(f'Resultados usando {labels[i]}')

  clf.fit(X_train, y_train.ravel())
 
  # predecir
  y_pred = clf.predict(X_test)

  # reporte de clasificación
  print('Matriz de confusión\n')
  print(confusion_matrix(y_test, y_pred))
  print('\nReporte de Clasificación\n')
  print(classification_report(y_test, y_pred))
  print('------------------------------------------------------------------')

Resultados usando Random Forest
Matriz de confusión

[[5935    0]
 [  57    8]]

Reporte de Clasificación

              precision    recall  f1-score   support

         0.0       0.99      1.00      1.00      5935
         1.0       1.00      0.12      0.22        65

    accuracy                           0.99      6000
   macro avg       1.00      0.56      0.61      6000
weighted avg       0.99      0.99      0.99      6000

------------------------------------------------------------------
Resultados usando Regresión Logística
Matriz de confusión

[[5935    0]
 [  50   15]]

Reporte de Clasificación

              precision    recall  f1-score   support

         0.0       0.99      1.00      1.00      5935
         1.0       1.00      0.23      0.38        65

    accuracy                           0.99      6000
   macro avg       1.00      0.62      0.69      6000
weighted avg       0.99      0.99      0.99      6000

-----------------------------------------------------------

Dado los resultados anteriores, estimamos conveniente elegir el modelo de Regresión Logística cambiando el parámetro `class_weight`, con el fin de solucionar el problema de desbalance de clases entre los clientes no fraudulentos y los que si lo son. Esta elección se hace observando la métrica Recall para la clase 1, pues es más importante para el banco es menos costoso equivocarse etiquetando como fraudelento a un cliento que no lo es, que etiquetar como no fraude a un cliente que si lo es, y esto es precisamente lo que la métrica Recall nos señala. 

### 6.2 Agregar resultados a dataframe de muestreo

In [37]:
# Se crea un dataframe auxiliar que corresponda a los mismos clientes que contiene X_test
df_sample_test = df_sample.loc[X_test.index]

In [38]:
# Se añade una columna de predicción a este dataframe
df_sample_test['prediction']=y_pred # Este y_pred corresponde a la regresión logística balanceada
df_sample_test['prediction']=df_sample_test['prediction'].astype(str) #Se cambia a string para poder visualizar de mejor manera los labels

### 6.3 Visualizar según rangos etarios

In [39]:
# Gráfico en dimensión reducida por segmentación entaria separado por fraude.  No Fraude = 0, Fraude = 1.
for edad in df_sample_test['segmentacion_etaria'].unique():
  fig = px.scatter(
    df_sample_test[df_sample_test['segmentacion_etaria']==edad],
    x="x_umap",
    y="y_umap",
    color='prediction',
    hover_data=['credit_risk_score' , 'phone_mobile_valid', 'fraud_bool']
  ) 

  # Cambiar dimensiones y título
  fig.update_layout(height=700, width=800, title_text="UMAP para "+edad)
  # Mostrar figura
  fig.show()

### 6.4 Calcular ratios y responder

In [40]:
total_fraude = len(df_sample_test[df_sample_test['prediction']=='1.0'])
tp_fraude = len(df_sample_test[(df_sample_test['prediction']=='1.0')&(df_sample_test['fraud_bool']=='1.0')])

P = tp_fraude/total_fraude
print('P =', P)

P = 0.28346456692913385


In [41]:
total_real_fraude = len(df_sample_test[df_sample_test['fraud_bool']=='1.0'])

R = tp_fraude/total_real_fraude
print('R =', R)

R = 0.5538461538461539


**¿Qué significan los ratios y sus valores?**

Los ratios P y R significan precision y recall respectivamente. 

La precisión se define como la proporción de instancias positivas clasificadas correctamente sobre el total de instancias clasificadas como positivas. 
$$Precisión = Verdaderos \ positivos / (Verdaderos \ positivos + Falsos \ positivos)$$

El recall mide la proporción de instancias positivas clasificadas correctamente sobre el total de instancias que realmente son positivas.

$$Recall = Verdaderos \ positivos / (Verdaderos \ positivos + Falsos \ negativos)$$

Estos valores van entre $0$ y $1$, y entre más alto sea su valor, mejor es el resultado de la predicción.

<br>

**¿Qué tan correctas fueron las predicciones realizadas por su modelo según los ratios calculados?**

Se obtuvieron métricas muy altas para la clase 0 (No fraude) debido al desbalance de los datos, lo cual produce un sesgo en el modelo que tiende a clasificar a la gran mayoría de clientes como no fraudulentos. Por otro lado, las métricas para la clase 1 (Fraude) son bastante deficientes, ya que hay una falta de ejemplos suficientes y, entre los disponibles, no se pudo identificar un patrón o correlación directa que permita predecir cuándo un cliente va a cometer fraude contra el banco.

<br>

**¿Son coherentes los resultados obtenidos?.**

A pesar de las métricas obtenidas, los resultados son coherentes pues el clasificador predice que la gran mayoría de clientes no son fraudulentos, lo que es cierto, y como se puede observar en los gráficos de dispersión realizados en la sección 5, un cliente fradulento puede compartir las mismas características que el resto.

<br>

**¿Cómo se comportan los casos de fraudes para los diferentes rangos etarios?**
Como se puede visualizar en los gráficos de dispersión del muestreo de datos utilizados, los clientes jóvenes no son fraudulentos. Esto se puede deber a que no tienen tanta relación con los bancos, y estos no suelen pedir prestamos bancarios. 

Otra observación es que a medida que aumenta la edad del cliente, las probabilidades de que sea fraudulento aumenta, siendo los adultos los que presentan mayor cantidad de personas que cometieron algún fraude. 

Sin embargo, para cada rango etario no se puede observar alguna característica en común que tengan los clientes fraudulentos en la proyección 2D realizada, sino que pararecen estar uniformemente distribuidos. Así que por esta razones suponemos que el modelo no es capaz de realizar buenas predicciones para abordar esta problemática. 